### PR TITLE
[Intel MKL] Compilation fixes for DNNL 1.0

### PR DESCRIPTION
**Pull request by nhasabni**
This PR makes following changes:

1. Enables MklLayoutPass for DNNL 1.0 integration
2. Makes certain changes to MKL kernels to fix compilation issues introduced by DNNL 1.0
3. Ports remaining features such as weight caching to DNNL 1.0 API
4. Adds `exponential_avg_factor` attribute to MklFusedBatchNorm ops
5. Adds BUILD support for DNNL 1.0

**NOTE**: Changes in `mkl_fused_batchnorm_op.cc` file reflect most of the changes made by PR #36767. Once this PR is merged to master, most of the changes in this file would go away.

### CODE REVIEW

**Comment by nhasabni** [comment]
Hi @penpornk, pls take a look whenever you get chance.

File: tensorflow/core/util/mkl_types.h [hunk 1]
```diff
@@ -112,6 +115,7 @@ namespace tensorflow {
 #define TENSOR_FORMAT_NHWC MKL_TENSOR_FORMAT_NHWC
 #define TENSOR_MAX_DIMS MKLDNN_MAX_NDIMS
 #define GET_USR_MEM_PRIM_DESC(src) src.GetUsrMemDesc()
+#define BN_FLAGS mkldnn::normalization_flags
```
Comment (penpornk) [inline tensorflow/core/util/mkl_types.h]: This is redundant with line 26. Please delete one. (And please keep the one for v0 below in the same relative order.)

File: tensorflow/core/kernels/mkl_matmul_ops_common.h [hunk 2]
```diff
@@ -438,49 +437,61 @@ class MklDnnMatMulOpBase : public OpKernel {
...
     // reorder and cache the weight
     weight.SetUsrMem(weight_md, &weight_tensor);
-    weight.CheckReorderToOpMem(matmul_fwd_pd.get()->weights_primitive_desc());
+    weight.CheckReorderToOpMem(MEMORY_PD_WITHOUT_DATA(
+        matmul_fwd_pd.get()->PRIMITIVE_DESC_WEIGHTS, cpu_engine_));
     weight_data = static_cast<Tweight*>(weight.GetOpMem().get_data_handle());
 
     Tensor* weight_tensor_ptr = nullptr;
 
+    size_t size = matmul_fwd_pd.get()->PRIMITIVE_DESC_WEIGHTS.get_size();
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_matmul_ops_common.h]: Nit: can we make this weight_size?
```suggestion
    size_t weight_size = matmul_fwd_pd.get()->PRIMITIVE_DESC_WEIGHTS.get_size();
```

File: third_party/mkl_dnn/mkldnn_v1.BUILD [hunk 3]
```diff
@@ -0,0 +1,135 @@
+exports_files(["LICENSE"])
```
Comment (penpornk) [inline third_party/mkl_dnn/mkldnn_v1.BUILD]: Could you please explain we why need a seperate file? (I'm not against it. I think it's good for safety. Just curious.)

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR! I have two minor changes requested and a few comments. I haven't looked at the files that overlap with PR #36767 yet. Will do after conflicts are merged.

File: tensorflow/core/kernels/mkl_matmul_ops_common.h (hunk 2 above)
Comment (nhasabni) [reply tensorflow/core/kernels/mkl_matmul_ops_common.h]: Thanks for reviewing, Penporn! Will address it.

File: third_party/mkl_dnn/mkldnn_v1.BUILD (hunk 3 above)
Comment (nhasabni) [reply third_party/mkl_dnn/mkldnn_v1.BUILD]: We thought about adding it to same mkldnn.BUILD file, but there were a lot of conditional checks that were complicating dependencies. And we would remove dnnl0.x sometime in the future. So having a separate file would make removal easy.

File: tensorflow/core/kernels/mkl_matmul_ops_common.h (hunk 2 above)
Comment (nhasabni) [reply tensorflow/core/kernels/mkl_matmul_ops_common.h]: Done. Pls take a look.

File: tensorflow/core/util/mkl_types.h (hunk 1 above)
Comment (nhasabni) [reply tensorflow/core/util/mkl_types.h]: Done. Pls take a look.

File: tensorflow/core/kernels/mkl_matmul_op_fused.cc [hunk 4]
```diff
@@ -86,11 +86,10 @@ class MklFusedMatMulOp : public MklDnnMatMulOpBase<T, T> {
...
 
-    OP_REQUIRES(
-        ctx, k == weight_tf_shape.dim_size(dim_pair[1]),
-        errors::InvalidArgument(
-            "Matrix size-incompatible: In[0]: ", src_tf_shape.DebugString(),
-            ", In[1]: ", weight_tf_shape.DebugString()));
+    OP_REQUIRES(ctx, k == weight_tf_shape.dim_size(dim_pair[1]),
+                errors::InvalidArgument("Matrix size-incompatible: In[0]: ",
+                                        src_tf_shape.DebugString(), ", In[1]: ",
+                                        weight_tf_shape.DebugString()));
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_matmul_op_fused.cc]: This reformatting shouldn't happen. Please revert.

File: tensorflow/core/kernels/mkl_matmul_op_fused.cc [hunk 5]
```diff
@@ -201,9 +200,9 @@ class MklFusedMatMulOp : public MklDnnMatMulOpBase<T, T> {
       // Execute fused matmul op.
       matmul_prim->Execute(src_data, weight_data, bias_data, dst_data);
     } catch (mkldnn::error& e) {
-      string error_msg = "Status: " + std::to_string(e.status) +
-                         ", message: " + string(e.message) + ", in file " +
-                         string(__FILE__) + ":" + std::to_string(__LINE__);
+      string error_msg = "Status: " + std::to_string(e.status) + ", message: " +
+                         string(e.message) + ", in file " + string(__FILE__) +
+                         ":" + std::to_string(__LINE__);
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_matmul_op_fused.cc]: This reformatting shouldn't happen. Please revert.

File: tensorflow/core/kernels/mkl_fused_batch_norm_op.cc [hunk 6]
```diff
@@ -14,14 +14,17 @@ limitations under the License.
...
 #include "tensorflow/core/framework/register_types.h"
 #include "tensorflow/core/framework/tensor.h"
 #include "tensorflow/core/framework/tensor_types.h"
 #include "tensorflow/core/util/mkl_types.h"
 #include "tensorflow/core/util/mkl_util.h"
 #include "tensorflow/core/util/tensor_format.h"
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
+
+#define GET_FLAG(bn_flag) static_cast<int>(BN_FLAGS::bn_flag)
+#define IS_SET(cflag) (context_.flags & GET_FLAG(cflag))
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_fused_batch_norm_op.cc]: This is redundant to lines 29-30. Please remove.

File: tensorflow/core/kernels/mkl_fused_batch_norm_op.cc [hunk 7]
```diff
@@ -432,12 +469,9 @@ class MklFusedBatchNormBwdPrimitive : public MklPrimitive {
...
 #ifdef ENABLE_MKLDNN_V1
     // Execute backward batch-normalization primitives.
-    // TODO(intel-tf): Use execute_primitive instead of the inlined code.
     DCHECK_EQ(context_.bwd_primitives.size(), context_.net_args.size());
-    for (size_t i = 0; i < context_.bwd_primitives.size(); ++i) {
-      context_.bwd_primitives.at(i).execute(*context_.bwd_stream,
-                                            context_.net_args.at(i));
-    }
+    execute_primitives(context_.bwd_primitives, context_.bwd_stream,
+                       context_.net_args);
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_fused_batch_norm_op.cc]: Thank you for the clean-up!

File: third_party/mkl_dnn/mkldnn_v1.BUILD (hunk 3 above)
Comment (penpornk) [reply third_party/mkl_dnn/mkldnn_v1.BUILD]: Makes sense. Thank you!

**Review by penpornk — CHANGES_REQUESTED** [review]
There are two reformattings that shouldn't  happen. Are you using `clang-format` 8.0.1?

**Comment by nhasabni** [comment]
> There are two reformattings that shouldn't happen. Are you using `clang-format` 8.0.1?

Yes, we are using clang+llvm-8.0.1. These changes were suggested by that tool.

**Comment by penpornk** [comment]
Got it. I'll try to find the version that works some time later then T-T. Please revert by hand for now.

File: tensorflow/core/kernels/mkl_matmul_op_fused.cc (hunk 4 above)
Comment (nhasabni) [reply tensorflow/core/kernels/mkl_matmul_op_fused.cc]: Done.

File: tensorflow/core/kernels/mkl_matmul_op_fused.cc (hunk 5 above)
Comment (nhasabni) [reply tensorflow/core/kernels/mkl_matmul_op_fused.cc]: Done.

File: tensorflow/core/kernels/mkl_fused_batch_norm_op.cc (hunk 6 above)
Comment (nhasabni) [reply tensorflow/core/kernels/mkl_fused_batch_norm_op.cc]: Done.

File: .bazelrc [hunk 8]
```diff
@@ -136,6 +136,7 @@ build --host_java_toolchain=//third_party/toolchains/java:tf_java_toolchain
 # environment variable "TF_MKL_ROOT" every time before build.
 build:mkl --define=build_with_mkl=true --define=enable_mkl=true
 build:mkl --define=tensorflow_mkldnn_contraction_kernel=0
+build:mkl --define=build_with_mkl_dnn_v1_only=true
```
Comment (penpornk) [inline .bazelrc]: Now that we have time, can we separate this to another PR? Just so it's cleaner/easier for debugging.

**Review by penpornk — CHANGES_REQUESTED** [review]
One last nit, please :)

File: .bazelrc (hunk 8 above)
Comment (nhasabni) [reply .bazelrc]: @penpornk should I submit separate PR now? or wait for this PR to approve and merge?

**Review by penpornk — APPROVED** [review]
Thank you for your patience! I think you can submit the other PR now. I'll make sure it's merged after this PR is merged.

**Comment by nhasabni** [comment]
@penpornk Thanks for quick review and approval. I've submitted another PR #37106 for making DNNL 1.2 default.

**Pull request merged**